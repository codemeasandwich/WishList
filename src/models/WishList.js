import { types } from "mobx-state-tree"

const data = {
  "name":"Cats DVD",
  "price":12.95,
  "image":"http://thecatapi.com/?id=4jm"
}

export const WishListItem = types.model({
  name:  types.string,
  price: types.number,
  image: ""
})

export const WishList = types.model({
  items: types.optional(types.array(WishListItem),[])
})