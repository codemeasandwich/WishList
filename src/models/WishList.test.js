import { getSnapshot, onSnapshot, onPatch } from "mobx-state-tree"
import { WishList, WishListItem } from "./WishList"

it("can create a instance of a model", () =>{
  const item = WishListItem.create({
      "name":"Cats DVD",
      "price":12.95
    })
  
  expect(item.price).toBe(12.95)
  expect(item.image).toBe("")
  item.changeName("Best of Cats")
  expect(item.name).toBe("Best of Cats")
})

it("can create a WishList", () =>{
  const list = WishList.create({items:[{
      "name":"Cats DVD",
      "price":12.95
    }]})
  
  expect(list.items.length).toBe(1)
  expect(list.items[0].price).toBe(12.95)
})

it("can add new items a WishList", () =>{
  const list = WishList.create();
  const states = [];
  
  onSnapshot(list, snapshot => {
    states.push(snapshot)  
  })
  
  const item = {
    "name":"Dog VHS",
    "price":2.45
  }
  list.add(item)
  
  expect(list.items.length).toBe(1)
  expect(list.items[0].name).toBe("Dog VHS")
  list.items[0].changeName("The Dog Video")
  expect(list.items[0].name).toBe("The Dog Video")
  
  expect(getSnapshot(list)).toMatchSnapshot() 
  expect(states).toMatchSnapshot() 
})

it("can add new items a WishList - 2", () =>{
  const list = WishList.create();
  const patchs = [];
  
  onPatch(list, patch => {
    patchs.push(patch)  
  })
  
  const item = {
    "name":"Dog VHS",
    "price":2.45
  }
  list.add(item)
  
  list.items[0].changeName("The Dog Video")
  
  expect(patchs).toMatchSnapshot() 
})