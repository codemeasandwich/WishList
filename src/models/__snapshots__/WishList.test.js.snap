// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add new items a WishList - 2 1`] = `
Array [
  Object {
    "op": "add",
    "path": "/items/0",
    "value": Object {
      "image": "",
      "name": "Dog VHS",
      "price": 2.45,
    },
  },
  Object {
    "op": "replace",
    "path": "/items/0/name",
    "value": "The Dog Video",
  },
]
`;

exports[`can add new items a WishList 1`] = `
Object {
  "items": Array [
    Object {
      "image": "",
      "name": "The Dog Video",
      "price": 2.45,
    },
  ],
}
`;

exports[`can add new items a WishList 2`] = `
Array [
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "Dog VHS",
        "price": 2.45,
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "The Dog Video",
        "price": 2.45,
      },
    ],
  },
]
`;
